irksome/imex.py

```diff
@@ -0,0 +1,267 @@
+    V = v.function_space()
+    Vbig = UU.function_space()
+    VV = TestFunction(Vbig)
+
+    num_stages = butch.num_stages
+    num_fields = len(V)
+    vc = np.vectorize(Constant)
+    Aexp = explicit_coeffs(num_stages)
+    Aprop = vc(Aexp)
+    Ait = vc(butch.A)
+    C = vc(butch.c)
+
+    u0bits, vbits, VVbits, UUbits = getBits(num_stages, num_fields,
+                                            u0, UU, v, VV)
+
+    Fit = Zero()
+    Fprop = Zero()
+
+    if splitting == AI:
+        for i in range(num_stages):
+            # replace test function
+            repl = {}
+
+            for k in range(num_fields):
+                repl[vbits[k]] = VVbits[i][k]
+                for ii in np.ndindex(vbits[k].ufl_shape):
+                    repl[vbits[k][ii]] = VVbits[i][k][ii]
+
+            Ftmp = replace(Fexp, repl)
+
+            # replace the solution with stage values
+            for j in range(num_stages):
+                repl = {t: t + C[j] * dt}
+
+                for k in range(num_fields):
+                    repl[u0bits[k]] = UUbits[j][k]
+                    for ii in np.ndindex(u0bits[k].ufl_shape):
+                        repl[u0bits[k][ii]] = UUbits[j][k][ii]
+
+                # and sum the contribution
+                replF = replace(Ftmp, repl)
+                Fit += Ait[i, j] * dt * replF
+                Fprop += Aprop[i, j] * dt * replF
+    elif splitting == IA:
+        # diagonal contribution to iterator
+        for i in range(num_stages):
+            repl = {t: t+C[i]*dt}
+            for j in range(num_fields):
+                repl[u0bits[j]] = UUbits[i][j]
+                repl[vbits[j]] = VVbits[i][j]
+
+            # Also get replacements right for indexing.
+            for j in range(num_fields):
+                for ii in np.ndindex(u0bits[j].ufl_shape):
+                    repl[u0bits[j][ii]] = UUbits[i][j][ii]
+                    repl[vbits[j][ii]] = VVbits[i][j][ii]
+
+            Fit += dt * replace(Fexp, repl)
+
+        # dense contribution to propagator
+        Ablah = vc(np.linalg.solve(butch.A, Aexp))
+
+        for i in range(num_stages):
+            # replace test function
+            repl = {}
+
+            for k in range(num_fields):
+                repl[vbits[k]] = VVbits[i][k]
+                for ii in np.ndindex(vbits[k].ufl_shape):
+                    repl[vbits[k][ii]] = VVbits[i][k][ii]
+
+            Ftmp = replace(Fexp, repl)
+
+            # replace the solution with stage values
+            for j in range(num_stages):
+                repl = {t: t + C[j] * dt}
+
+                for k in range(num_fields):
+                    repl[u0bits[k]] = UUbits[j][k]
+                    for ii in np.ndindex(u0bits[k].ufl_shape):
+                        repl[u0bits[k][ii]] = UUbits[j][k][ii]
+
+                # and sum the contribution
+                Fprop += Ablah[i, j] * dt * replace(Ftmp, repl)
+    else:
+        raise NotImplementedError("Unsupported splitting type")
```

Comment:
It seems odd that we default to `splitting = None`, but then the error is "Unsupported splitting type".  Maybe this could be "Must specify splitting to either IA or AI"?

Reply:
This makes sense.